salt/_utils/kubernetes_utils.py

```diff
@@ -0,0 +1,535 @@
+    def __init__(self, cr_details, manifest):
+        self._details = cr_details
+        self._body = _cast_dict_keys(manifest, _camel_to_snake)
+        self._attr_dict = _DictWrapper(self._body)
+
+    details = property(operator.attrgetter('_details'))
+    body = property(operator.attrgetter('_body'))
+
+    def to_dict(self):
+        return self.body
+
+    def to_str(self):
+        return pformat(self.to_dict())
+
+    def __repr__(self):
+        return self.to_str()
+
+    def __eq__(self, other):
+        if not isinstance(other, CustomObject):
+            return False
+
+        return self.body == other.body
+
+    def __getattribute__(self, name):
+        try:
+            return super(CustomObject, self).__getattribute__(name)
+        except AttributeError:
+            return getattr(self._attr_dict, name)
+
+
+def __virtual__():
+    return __virtualname__
+
+
+def get_object_details(manifest):
+    try:
+        api_version = manifest['apiVersion']
+        kind = manifest['kind']
+    except KeyError:
+        raise ValueError(
+            'Make sure to provide a valid Kubernetes manifest, including'
+            ' `kind` and `apiVersion` fields.'
+        )
+
+    object_details = KNOWN_CUSTOM_OBJECTS.get(
+        (api_version, kind),
+        KNOWN_STD_OBJECTS.get((api_version, kind))
+    )
+
+    if object_details is None:
+        raise ValueError(
+            'Unknown object type provided: {}/{}. Make sure that it is'
+            ' registered properly.'.format(api_version, kind)
+        )
+
+    return object_details
+
+
+def convert_manifest_into_object(manifest):
+    """Convert a YAML representation of a K8s object into its Python model.
+
+    This method can only convert known types, as declared in the constants
+    `KNOWN_STD_OBJECTS` or `KNOWN_CUSTOM_OBJECTS` in this module.
+
+    It will also create the Python objects from sub-dicts when necessary before
+    building the final result.
+    """
+    object_details = get_object_details(manifest)
+
+    if isinstance(object_details, CRDetails):
+        return CustomObject(object_details, manifest)
+
+    return _build_standard_object(object_details.model, manifest)
+
+
+def _build_standard_object(model, manifest):
+    reverse_attr_map = {
+        value: key for key, value in model.attribute_map.items()
+    }
+
+    kwargs = {}
+    for src_key, src_value in manifest.items():
+        key = reverse_attr_map.get(src_key, src_key)
+        type_str = model.swagger_types.get(key)
+
+        if type_str is None:
+            raise ValueError(
+                'Unsupported attribute {} for "{}" object.'.format(
+                    src_key, model.__name__
+                )
+            )
+
+        try:
+            value = _cast_value(src_value, type_str)
+        except TypeError as exc:
+            raise ValueError(
+                'Invalid type for attribute {} of a "{}" object: {}.'.format(
```

Comment:
Invalid value rather than type, no?

Reply:
Yes, the mention of wrong type would be in `str(exc)` anyway.